### DISCUSSION TITLE
How to install NVIDIA packages (CUDA, tensorrt... )

### DISCUSSION THREAD

**Discussion by CHNtentes** [OE4T/meta-tegra#1114] [Q&A] [answered]
Hello,

I came to this github repo from Ridgerun guide [https://developer.ridgerun.com/wiki/index.php/Yocto_Support_for_NVIDIA_Jetson_Platforms/](url) . Although I'm able to build the image, adding NV packages brings some issues for me:

1. In the guide, it tells me to use `IMAGE_INSTALL_append = " cuda-samples"` to add CUDA package. However, under _/usr/local/cuda-11.4/bin_, there's nothing but only _cuda-install-samples-11.4.sh_. Obviously, I expect some binary files in this directory. 
2. When I add `IMAGE_INSTALL_append = " tensorrt"` in _local.conf_, bitbake returns _ERROR: Nothing RPROVIDES 'tensorrt'_. I tried `IMAGE_INSTALL_append = " tensorrt-samples"` and no error occurs, but I doubt this is the solution here.

I'm using _kirkstone_ branch for all layers and _Orin AGX devkit_. Thanks for support.

**Comment by madisongh** [comment, marked as answer]
For `cuda-samples`, by reading the [recipe](https://github.com/OE4T/meta-tegra/blob/kirkstone/recipes-devtools/cuda/cuda-samples_11.4.239-1.bb) you'll see that it builds only a couple of the sample programs, and they [get installed](https://github.com/OE4T/meta-tegra/blob/9d58efcfad33a0496b2099e01111c4c937bf3287/recipes-devtools/cuda/cuda-samples_11.4.239-1.bb#L92) in `/usr/bin/cuda-samples/`.

For TensorRT, I suspect that the wiki page is just a bit out of date.  To provide more control over exactly what packages get installed, and to maximize the number of packages that get built from sources, there are different recipes for the different bits of TensorRT that you might need.  Take a look at [the recipes](https://github.com/OE4T/meta-tegra/tree/kirkstone/recipes-devtools/gie) for more information.

Unlike JetPack, which dumps a huge blob containing as much of NVIDIA's software on your device as possible so you can play with it, the goal here is to let you pick and choose exactly what you need for your product/application. Because of that basic difference, you won't find a direct 1:1 mapping between JetPack packages and those generated by the recipes here.

**Reply by leadercxn** [reply]
Hi @madisongh ,
As your mean, I install cudnn, libnvvpi1, libvisionworks,  tensorrt-core ...etc packages
<img width="585" alt="image" src="https://user-images.githubusercontent.com/43691182/219262166-eedbe7e0-86a6-4279-9393-3de806f075c5.png">
But I run `jtop` and watch the system info, it shows these packages `NOT_INSTALLED`
<img width="644" alt="企业微信截图_8f8c4e39-0abd-4404-82eb-9465aea58f6e" src="https://user-images.githubusercontent.com/43691182/219262811-1e0d8e6a-cea9-493c-a8c2-050e7ca24029.png">
In fact, when I ran the demo, it was fine.  So this is a `python3-jetson-stats` bug ?
<img width="960" alt="image" src="https://user-images.githubusercontent.com/43691182/219262904-b6f51d3a-9e7d-4434-9395-77e2f55bf0fb.png">

**Reply by madisongh** [reply]
>  So this is a python3-jetson-stats bug ?

If that's the package that provides `jtop`, then probably so.  I've never used that tool, or even looked at it, but I'm guessing it's assuming it's running on NVIDIA's Ubuntu distro and looking for installed packages using techniques that work there but would be completely wrong for any other OS.  I'm not sure the maintainer of that package would consider that a bug, though.

**Reply by ichergui** [reply]
> So this is a python3-jetson-stats bug ?

I introduced the recipe at some point. BTW, it is not up to date.
I didn't try it with the latest BSP. I need to double check. 
IIRC, I tried it with L4T R32.X.Y and it was working fine.

**Reply by ichergui** [reply]
Hey @leadercxn 
Please try our latest changes [#63](https://github.com/OE4T/meta-tegra-community/pull/63) which includes `jtop` fixes
Let us know if you notice any issue

**Reply by leadercxn** [reply]
> Hey @leadercxn Please try our latest changes [#63](https://github.com/OE4T/meta-tegra-community/pull/63) which includes `jtop` fixes Let us know if you notice any issue

Okay, I will test it and let you know the result.thanks~

**Reply by ichergui** [reply]
Hey @leadercxn 
Any updates ?

**Reply by leadercxn** [reply]
> Hey @leadercxn Any updates ?

@ichergui  Yes, I copied the bb files and directories from the latest `master` branch and used them.
<img width="504" alt="image" src="https://user-images.githubusercontent.com/43691182/221085691-e572320d-cca8-4165-ac73-949782fd9823.png">

When I used `jtop` again, it was better than before
![wecom-temp-209323-6e6fdf4e116bcb6987b267072b4b6587](https://user-images.githubusercontent.com/43691182/221085562-3e4058d1-9b17-48df-9f83-2d48501c0ae8.jpg)

Thank you ~

**Reply by leadercxn** [reply]
> Hey @leadercxn Any updates ?

Will you sync to the `kirkstone-l4t-r32.7.x` branch of the `meta-tegra-community` repository ?

**Reply by ichergui** [reply]
Yes, I will do it over the weekend

**Reply by ichergui** [reply]
Hey @leadercxn 

The branch `kirkstone-l4t-r32.7.x` from `meta-tegra-community` repo is updated

**Comment by dwalkes** [comment]
@CHNtentes if you haven't noticed it already you may want to look at the repo at https://github.com/OE4T/tegra-demo-distro which is a fully working build setup with some customized images.  The demo images there [include a packagegroup-demo-basetests](https://github.com/OE4T/tegra-demo-distro/blob/a6f55265c402a5affe1409abf2340d53b00201ec/layers/meta-tegrademo/recipes-demo/images/demo-image-common.inc#L7) which [includes cuda-samples](https://github.com/OE4T/tegra-demo-distro/blob/a6f55265c402a5affe1409abf2340d53b00201ec/layers/meta-tegrademo/recipes-demo/packagegroups/packagegroup-demo-basetests.bb#L8).  Regarding how these images are tested you can see previous release tests and test results in [this spreadsheet](https://docs.google.com/spreadsheets/d/1EWG5QZvtWdk74gz48D4xCbegK-bsqMFJ8XOsyGdUUeI/edit).  See the CUDA tests there for the tests we currently have for cuda-samples.  Contributions in terms of running or adding tests are welcome!
